OpenKeychain/src/main/java/org/sufficientlysecure/keychain/ssh/SSHAuthenticationOperation.java

```diff
@@ -0,0 +1,296 @@
+import org.sufficientlysecure.keychain.pgp.Progressable;
+import org.sufficientlysecure.keychain.pgp.exception.PgpGeneralException;
+import org.sufficientlysecure.keychain.pgp.exception.PgpKeyNotFoundException;
+import org.sufficientlysecure.keychain.provider.KeyRepository;
+import org.sufficientlysecure.keychain.service.input.CryptoInputParcel;
+import org.sufficientlysecure.keychain.service.input.RequiredInputParcel;
+import org.sufficientlysecure.keychain.util.Log;
+import org.sufficientlysecure.keychain.util.Passphrase;
+
+import java.util.Collection;
+
+/**
+ * This class supports a single, low-level, authentication operation.
+ * <p/>
+ * The operation of this class takes a SSHAuthenticationParcel
+ * as input, and signs the included challenge as parametrized in the
+ * SSHAuthenticationData object. It returns its status
+ * and a possible signature as a SSHAuthenticationResult.
+ * <p/>
+ *
+ * @see SSHAuthenticationParcel
+ */
+public class SSHAuthenticationOperation extends BaseOperation<SSHAuthenticationParcel> {
+
+    private static final String TAG = "SSHAuthenticationOperation";
+    private static final String HASHALGORITHM = "SHA1";
+    private static final int HASHALGORITHM_NONE = 0;
+
+
+    public SSHAuthenticationOperation(Context context, KeyRepository keyRepository, Progressable progressable) {
+        super(context, keyRepository, progressable);
+    }
+
+    @NonNull
+    @Override
+    public SSHAuthenticationResult execute(SSHAuthenticationParcel input,
+                                           CryptoInputParcel cryptoInput) {
+        OperationLog log = new OperationLog();
+        SSHAuthenticationResult result = executeInternal(input.getAuthenticationData(), cryptoInput, input);
+        return result;
+    }
+
+    @NonNull
+    public SSHAuthenticationResult execute(SSHAuthenticationData data,
+                                           CryptoInputParcel cryptoInput,
+                                           SSHAuthenticationParcel sshAuthenticationParcel) {
+        return executeInternal(data, cryptoInput, sshAuthenticationParcel);
+    }
+
+    /**
+     * Signs challenge based on given parameters
+     */
+    private SSHAuthenticationResult executeInternal(SSHAuthenticationData data,
+                                                    CryptoInputParcel cryptoInput,
+                                                    SSHAuthenticationParcel sshAuthenticationParcel) {
+        int indent = 0;
+        OperationLog log = new OperationLog();
+
+        log.add(LogType.MSG_PSE, indent);
+        indent += 1;
+
+        Log.d(TAG, data.toString());
+
+        long opTime = 0;
+        long startTime = System.currentTimeMillis();
+
+        updateProgress(R.string.progress_extracting_signature_key, 0, 100);
+
+        byte[] signature = null;
+
+        byte[] challenge = sshAuthenticationParcel.getChallenge();
+        if (challenge == null) {
+            Log.d(TAG, "No challenge!");
```

Comment:
AuthenticationParcel.getChallenge is `@Nullable` (since the getters in PgpSignEncryptInputParcel are too and PgpSignEncryptOperation handles these errors; I used those as reference)

Reply:
If null is not a useful input value, why is it nullable? :)

Reply:
Good point, I will move the error handling for the challenge to SSHAgentService